### ISSUE TITLE
Safekeeper <-> pageserver connection shutdown

### ISSUE THREAD

**Issue by kelvich** [neondatabase/neon#1074]
Recently we've implemented logic to shutdown connection between safekeeper and pageserver when compute is down. Right after that we discovered that such behavior can lead to a distributed livelock:
1. Compute went down
2. Safekeeper immediately closed connection to the pageserver, but some WAL left unreplicated
3. Compute starts and asks a basebackup@latest_lsn from the pageserver 
4. Pageserver never saw `latest_lsn` so it waits for it. But safekeeper has no intention to replicate that WAL before compute will start a new connection.

So far we are addressing that issue by letting safekeeper finish its replication before shutting down. But I can see at least two other cases that are not covered with this approach:

1. During that residual replication safekeeper restarts (e.g. due to deploy or some random reason). Then it still would have some unreplicated data.
2. After that residual replication is done pageserver restarts/crashes and it's last_received_lsn decreases. So now we are again in the same situation -- even that some time ago everything was fine.

cc @arssher @lubennikovaav @petuhovskiy

**Comment by arssher** [comment]
I assumed persisting (latest) pageserver address in safekeeper to deal with 1. That doesn't cover 2 though, hm.

**Comment by arssher** [comment]
A couple of more options:
- Wait on safekeeper side for actual checkpoint happening on pageserver before putting timeline offline, i.e. disk_consistent_lsn must reach commit_lsn. Even if there isn't much data to checkpoint, after reasonable amount of time this should happen.
- (less attractive) When during compute creation basebackup errors out with 'timed out waiting for WAL', console can ping safekeepers 'please retry push to pageserver', or just restart compute creation -- such ping can be included in sync-safekeepers anyway. And note that if it is included, then 2 becomes a problem only if pageserver is restarted between sync-safekeepers and basebackup -- the probability of this is quite low.

**Comment by arssher** [comment]
For the record, it is clear `--sync-safekeepers` significantly makes probability of this issue really small, as its connection revives the one to pageserver. However, it doesn't eliminate it completely, e.g. pageserver restart after sync-safekeepers and before basebackup might still lose latest WAL.

**Comment by shanyp** [comment]
@arssher is there any requirement from pageserver ? if not lets remove the pageserver label

**Comment by petuhovskiy** [comment]
This issue looks like it was created when we had `callmemaybe` (before `etcd` and `broker`), so it's no longer relevant. We may still have similar problems, but I think we can close this issue.

>  is there any requirement from pageserver ?

`walreceiver` is a part of pageserver, so it makes sense to use pageserver label when issue is related to the `walreceiver` code. Or do you have other suggestions?
